pixeltable/exprs/data_row.py

```diff
@@ -123,11 +123,18 @@ def __getitem__(self, index: object) -> Any:
         assert self.has_val[index], index
 
         if self.file_urls[index] is not None and index in self.img_slot_idxs:
-            # if we need to load this from a file, it should have been materialized locally
-            assert self.file_paths[index] is not None
-            if self.vals[index] is None:
-                self.vals[index] = PIL.Image.open(self.file_paths[index])
+            if self.file_urls[index].startswith('data:'):
+                # It's a base64-encoded data URL. Decode it now.
+                with urllib.request.urlopen(self.file_urls[index]) as response:
```

Comment:
Didn't this already get downloaded by the CachePrefetchNode and the file is available in file_paths?

Reply:
That’s what I expected as well, and it’s indeed the case when defining a computed column. But it fails with an assertion error if selecting the image cast dynamically without defining a new column, as is done in the example notebook. 

Maybe that’s a bug - and this probably needs more investigation. let’s discuss at standup Monday.

Reply:
Let's punt on that bug right now (and change the notebook not to trigger it). Doing the download during expr evaluation would have very poor performance at the moment.

We'll fix that later in a larger overhaul of expr evaluation and parallelization.

Reply:
Okay, so this isn't really a url, it's the data inlined. I didn't understand that before.

How does the data: url end up in here? I'm assuming TypeCast gets a string and then calls ImageType.create_literal()? It seems like that's where the conversion should happen. file_urls should stick with file urls.